### PR TITLE
framework for providing high availability for OVN DBs

### LINKED ISSUE #648: Provide framework to enable high availability of OVN DBs
We need a way to  run OVN DBs in highly available manner. Any solution that we are going to come up with should ensure that OVN DBs can be run either natively in the K8s cluster or outside of the K8s cluster.

For more information, please see the community commented document here:
https://docs.google.com/document/d/1khkgTNvjSEKgOSrzJOTtApL_mai-f3-pW1NHXOv5Dak/edit?usp=sharing

The idea here is to have 3 sets of PODs:

**ovnkube-db POD** -- runs both OVN NB DB and OVN SB DB. Other PODs will wait until this POD comes up since it has DBs. This will be facilitated using headless service without label selectors -- named ovnkube-db. The endpoint for this service will be populated by ovnkube.sh once both the DBs are up.

**ovnkube-master POD** -- runs ovn-northd and ovnkube in master mode. This POD depends on ovnkube-db and becomes aware of the availability of the DB POD by checking the ovnkube-db Endpoint object (just like how we do it today).

**ovnkube-node POD** -- runs ovs-servers, ovn-controller, and ovnkube in Node mode. This POD depends on ovnkube-db and becomes aware of the availability of the DB POD by checking the ovnkube-db Endpoint object (just like how we do it today).

### PR DESCRIPTION
**Pull request by girishmg**
The idea here is to have 3 sets of PODs:

1. ovnkube-db POD -- runs both OVN NB DB and OVN SB DB. Other PODs will wait
until this POD comes up since it has DBs.  This will be facilitated using
headless service without label selectors -- named ovnkube-db. The endpoint for
this service will be populated by ovnkube.sh once both the DBs are up.

2. ovnkube-master POD -- runs ovn-northd and ovnkube in master mode. This POD
depends on ovnkube-db and becomes aware of the availability of the DB POD by
checking the ovnkube-db Endpoint object (just like how we do it today).

3. ovnkube-node POD – no changes and will have containers like today.

Fixes #648

Signed-off-by: Girish Moodalbail <gmoodalbail@nvidia.com>
Signed-off-by: Yun Zhou <yunz@nvidia.com>

### CODE REVIEW

**Comment by girishmg** [comment]
@pecameron can you please take a look?

**Comment by girishmg** [comment]
> 
> In general adding a pod increases complexity and the corner cases that must be correctly handled and tested.

Yes, we have tested this to the extent possible. There shouldn't be any regression from what we already have today.

> 
> SOAP-BOX ON
> OVN today is a technology demo/preview, it works in a very limited set of conditions, you install it and you can create/delete pods. This is because at install time everything is in sync (there is nothing there). You can't restart any of the network components and expect it to still work. It is a long way from being a product capable of handling whatever happens.
> SOAPBOX-OFF

I agree with your SOAPBOX comment above. Hopefully, all the efforts we are putting together as a community will make OVN CNI one of the most reliable and scalable CNI out there.

Yes, we need a Sync thread that always Sync data from k8s etcd data store to OVN data store.

File: dist/images/ovnkube.sh [hunk 1]
```diff
@@ -147,17 +151,20 @@ wait_for_event () {
 
 }
 
+# OVN DBs must be up and initialized before ovn-master and ovn-node PODs can come up
```
[the remark this answers is not in the document]
Comment (girishmg) [reply dist/images/ovnkube.sh]: Yes, this syncing happens today in various sync*() methods when the onvkube-master comes up. Also, @dcbw added few changes where in we wait for logical switches  to come up before creating logical_switch_ports.

[the remark this answers is not in the document]
Comment (pecameron) [reply dist/images/ovnkube.sh]: When ovn db comes up and master comes up, how do you lockout node until master is done syncing?

**Comment by pecameron** [comment]
@girishmg my comment on testing was more in the spirit of there being more combinations to test when you add a new pod. Testing becomes much more difficult. There are more boundary conditions and restart sequences and races that complicates testing. The tests need to be in CI and they need to flush out races as well as test features and must be repeatable.

File: dist/images/ovnkube.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (girishmg) [reply dist/images/ovnkube.sh]: The ovnkube-node waits in a loop for the node annotation of ovn_host_subnet to appear. See StartClusterNode(). This particular annotation is added by ovnkube-master for that node after creating the LogicalSwitch for that node.

The flow of events on the master are following and it starts in StartClusterMaster()

-- it updates subnetAllocatorList with already allocated subnets for the existing K8s nodes
-- it sets up the logical topology -- ovn_cluster_router, join switch, and so on
-- it calls watchNodes(). This particular function calls syncNodes() first and then calls addNode() for each of the existing nodes. 
-- in addNode(), the logical_switch for the node is created and then ovn_host_subnet node annotation happens.

 So, ovnkube-node will not proceed until the last step above occurs.

**Comment by shettyg** [comment]
We don't have any CI now. It is a very large effort to do any opensource  CI. The best we can do is review and fix when we see issues.
